charms/kfp-metadata-writer/tests/unit/test_operator.py

```diff
@@ -0,0 +1,182 @@
+    # Arrange
+    harness.begin()
+
+    # Mock:
+    # * leadership_gate to be active and executed
+    harness.charm.leadership_gate.get_status = MagicMock(return_value=ActiveStatus())
+
+    # Add relation with data.  This should trigger a charm reconciliation due to relation-changed.
+    add_sdi_relation_to_harness(harness, "grpc", data=MOCK_GRPC_DATA)
+
+    # Assert
+    assert isinstance(harness.charm.grpc_relation.status, ActiveStatus)
+
+
+def test_grpc_relation_without_data(harness, mocked_lightkube_client):
+    """Test that the grpc relation goes Blocked if no data is available."""
+    # Arrange
+    harness.begin()
+
+    # Mock:
+    # * leadership_gate to be active and executed
+    harness.charm.leadership_gate.get_status = MagicMock(return_value=ActiveStatus())
+
+    # Add relation with data.  This should trigger a charm reconciliation due to relation-changed.
+    add_sdi_relation_to_harness(harness, "grpc", data={})
+
+    # Assert
+    assert isinstance(harness.charm.grpc_relation.status, BlockedStatus)
+
+
+def test_grpc_relation_without_relation(harness, mocked_lightkube_client):
+    """Test that the grpc relation goes Blocked if no relation is established."""
+    # Arrange
+    harness.begin()
+
+    # Mock:
+    # * leadership_gate to be active and executed
+    harness.charm.leadership_gate.get_status = MagicMock(return_value=ActiveStatus())
+
+    # Act
+    harness.charm.on.install.emit()
+
+    # Assert
+    assert isinstance(harness.charm.grpc_relation.status, BlockedStatus)
+
+
+def test_pebble_service_container_running(harness, mocked_lightkube_client):
+    """Test that the pebble service of the charm's kfp-metadata-writer container is running."""
+    harness.set_leader(True)
+    harness.begin()
+    harness.set_can_connect("kfp-metadata-writer", True)
+
+    harness.charm.kubernetes_resources.get_status = MagicMock(return_value=ActiveStatus())
+    add_sdi_relation_to_harness(harness, "grpc", data=MOCK_GRPC_DATA)
+
+    harness.charm.on.install.emit()
+
+    assert isinstance(harness.charm.unit.status, ActiveStatus)
+
+    container = harness.charm.unit.get_container("kfp-metadata-writer")
+    # Assert that sidecar container is up and its service is running
+    assert container.get_service("kfp-metadata-writer").is_running()
+
+    # Assert the environment variables that are set from defaults in the service plan
+    environment = container.get_plan().services["kfp-metadata-writer"].environment
+    assert environment["NAMESPACE_TO_WATCH"] == ""
+    assert (
+        environment["METADATA_GRPC_SERVICE_SERVICE_HOST"]
+        == harness.charm.grpc_relation.component.get_data()["service"]
+    )
+    assert (
+        environment["METADATA_GRPC_SERVICE_SERVICE_PORT"]
+        == harness.charm.grpc_relation.component.get_data()["port"]
+    )
+
+
+def test_pebble_service_is_replanned_on_config_changed(harness, mocked_lightkube_client):
```

Comment:
This test case doesn't seem like checking what you intend. Maybe what you want to do is emit a config changed, and then check that the `replan()` method is called, though it is embedded in the base charm.
On another note, is the expected behaviour of the charm to replan on config changed? I don't see anything that actually triggers that behaviour.

Reply:
You are right, I was confused from what we tested in other charms. I 'm removing this test as it is pointless.